src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
+        } else {
+            Node<T> current = findCurrent(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+    }
+
+    public Node<T> findCurrent(int index) {
+        validateIndex(index, size);
+        Node<T> current = (index < size / 2) ? head : tail;
+        if (index < size / 2) {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index, size);
+        return findCurrent(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index, size);
+        Node<T> current = findCurrent(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index, size);
+        Node<T> currentNode = findCurrent(index);
+        T removedValue = currentNode.value;
+        unlink(currentNode);
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.value == null)
+                    || (object != null && object.equals(current.value))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            head = head.next;
+            if (head != null) {
+                head.prev = null;
+            } else {
+                tail = null;
+            }
+        } else if (node == tail) {
+            tail = tail.prev;
+            if (tail != null) {
+                tail.next = null;
+            } else {
+                head = null;
+            }
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void validateIndex(int index, int upperBound) {
+        if (index < 0 || index >= upperBound) {
+            throw new IndexOutOfBoundsException(MESSAGE_OUT_OF_BOUNDS + ": " + index);
```

Comment:
The `validateIndex` method should throw `ArrayListIndexOutOfBoundsException` instead of `IndexOutOfBoundsException` to align with the custom exception handling in your project.